FLEDGE.md

```diff
@@ -202,6 +206,8 @@ The returned `auctionResultPromise` object is _opaque_: it is not possible for a
 
 Optionally, `sellerTimeout` can be specified to restrict the runtime (in milliseconds) of the seller's `scoreAd()` script, and `perBuyerTimeouts` can be specified to restrict the runtime (in milliseconds) of particular buyer's `generateBid()` scripts. If no value is specified for the seller or a particular buyer, a default timeout of 50 ms will be selected. Any timeout higher than 500 ms will be clamped to 500 ms. A key of `'*'` in `perBuyerTimeouts` is used to change the default of unspecified buyers.
 
+`perBuyerBidGenerationTimeouts` is structured like `perBuyerTimeouts`, but the values cover the entirety of the time it takes to generate bids for all interest groups for each buyer, including downloading resources, starting processes, and all generate bid calls. The single limit applies collectively across all interest groups with the same owner. It's measure as wall clock time starting when dedicated tasks needed to generate a bid for a particular buyer start. Timers may be running for multiple bidders simultaneously.  It does not include time taken by the seller to score the buyer's bids. Once the timer expires, the affected buyer's interest groups may no longer generate any bids. Scripts may be unloaded, fetches cancelled, etc. All bids generated before the timeout will continue to participate in the auction. FLEDGE implementations should attempt, on a best-effort basis, to generate bids for each buyer in priority order, so lower priority interest groups are the ones more likely to be timed out.
```

Comment:
> FLEDGE implementations should attempt, on a best-effort basis, to generate bids for each buyer in priority order

In the case of end user devices with a limited number of hardware threads, is there a recommendation on how FLEDGE implementations should schedule the CPU execution time _among_ individual buyers, given `perBuyerBidGenerationTimeouts` constraints are present and measure the wall clock time? For example, should the bidding process for different interest group buyers happen sequentially or in parallel (perhaps, relying on OS-provided scheduling), and how would that interact with the measurement of `perBuyerBidGenerationTimeouts`?

Reply:
> > FLEDGE implementations should attempt, on a best-effort basis, to generate bids for each buyer in priority order
> 
> In the case of end user devices with a limited number of hardware threads, is there a recommendation on how FLEDGE implementations should schedule the CPU execution time _among_ individual buyers, given `perBuyerBidGenerationTimeouts` constraints are present and measure the wall clock time? For example, should the bidding process for different interest group buyers happen sequentially or in parallel (perhaps, relying on OS-provided scheduling), and how would that interact with the measurement of `perBuyerBidGenerationTimeouts`?

Good question!

So Chrome for desktop's implementation loads and runs scripts for 10 buyers and 3 sellers at a time (which is arbitrary - number is more based on concerns about process overhead than CPU contention), loading buyers in the order they appear in the auction.

I'm not sure what, if anything, we'd want to add about specific details, though CPU contention and implementation details certainly come into consideration here.  I think it makes sense to recommend buyers be loaded in the order they appear in the auction - we generally want to put as much control in the hands of auction participants as we reasonably can.  Chrome does this, though if there's a FLEDGE process already running for a buyer, the buyer will jump the queue.  On Android, where we don't use a dedicated FLEDGE process, this behavior currently means all buyers typically run in parallel.

There's a tradeoff between speccing things out in more detail to force consistency across browsers/platforms in the spec, and allow browsers to optimize behavior given their own resource constraints here.

The wording here sort of assumes that scripts are loaded and then run on a per-buyer basis immediately.  Could imagine a browser implementation that downloaded all scripts at once, and then scheduled them based on CPU availability, in which case, these would effectively become timeouts for the entire bid generation phase of the auction.

Reply:
> loading buyers in the order they appear in the auction

If some form of a queue-like mechanism was employed by a browser due to resource constraints (such as # of threads or processes that can execute in parallel), would a timer for the given buyer only be started once that buyer's task(s) are removed from the start of the queue, and the buyer's `generateBid` functions received a chance to execute (as opposed to the time when these tasks had originally been added to the queue)?

Reply:
> > loading buyers in the order they appear in the auction
> 
> If some form of a queue-like mechanism was employed by a browser due to resource constraints (such as # of threads or processes that can execute in parallel), would a timer for the given buyer only be started once that buyer's task(s) are removed from the start of the queue, and the buyer's `generateBid` functions received a chance to execute (as opposed to the time when these tasks had originally been added to the queue)?

The intention is that it would start as soon as something is done specifically for the buyer that delays the auction, such as starting to download the JS/JSON (In Chrome's case, this would actually be starting the process to download the JS/JSON - Chrome then runs the JS as soon as it has downloaded all it needs, modulo platform CPU time queuing).

The intention is to try and keep this simple and allow for future speed improvements (e.g., if we had an API that set a time limit on the combined JS execution, but someone had a 32 core CPU and run a bunch of bidder scripts in parallel, odds are that it's wall clock time that matters to sellers, not the total time across all 32/64 threads all running bidder scripts at once)